File: LeetCode/5회차/B08-42-Trapping Rain Water2.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        res = 0
        stack = []

        for i in range(len(height)):
            while stack and height[stack[-1]] < height[i]:
                top = stack.pop()

                if not stack:
                    break

                water = min(height[i] - height[top], height[stack[-1]] - height[top])
                res += water * (i - stack[-1] - 1)

            stack.append(i)

        return res


print(Solution().trap([0,1,0,2,1,0,1,3,2,1,2,1]))

File: LeetCode/JavaScript/B08-42-Trapping Rain Water2.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
  stack = [];
  volume = 0;

  for (let i = 0; i < height.length; i++) {
    while (!!stack && height[stack[stack.length - 1]] < height[i]) {
      const top = stack.pop();
      if (stack.length === 0) {
        break;
      }
      const dist = i - stack[stack.length - 1] - 1;
      const water = Math.min(height[stack[stack.length - 1]], height[i]) - height[top];
      volume += dist * water;
    }
    stack.push(i);
  }

  return volume;
};

console.log(trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]));
